### PR TITLE
resolved notifications: follow-up fixes

### PR DESCRIPTION
**Pull request by bluca**
Further review comments from: https://github.com/systemd/systemd/pull/22845

### CODE REVIEW

**Comment by bluca** [comment]
See comment on data format: https://github.com/systemd/systemd/pull/22845#issuecomment-1242303923

File: src/resolve/resolved-dns-query.c [hunk 1]
```diff
@@ -587,8 +586,7 @@ void dns_query_complete(DnsQuery *q, DnsTransactionState state) {
         q->state = state;
 
         if (state == DNS_TRANSACTION_SUCCESS && set_size(q->manager->varlink_subscription) > 0) {
-                DnsQuestion *question = q->request_packet ? q->request_packet->question : NULL;
-                const char *query_name = question ? dns_question_first_name(question) : q->request_name;
+                const char *query_name = dns_question_first_name(q->request_packet ? q->request_packet->question : q->question_utf8);
```
Comment (bluca) [inline src/resolve/resolved-dns-query.c]: Check on centos (but not on ubuntu) is now failing and this is the only functional change, so I assume it's due to this.

Test is receiving `{"parameters":{"addresses":[{"ifindex":6,"family":2,"address":[10,0,0,14],"type":"A"}],"name":"follow10.so.close.signed.test"},"continues":true}` but it's expecting `cname-chain.signed.test` as the name. @mrc0mmand Any idea? This was changed following @poettering request

Comment (mrc0mmand) [reply src/resolve/resolved-dns-query.c]: No idea, unfortunately. I've also noticed there's a race condition somewhere where we might miss some notifications, as already happened in several unrelated PRs, e.g.: https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-centos8/11030/artifact//systemd-centos-ci/artifacts_wf1eehdn/testsuite-logs-upstream.bBa/TEST-75-RESOLVED_FAIL.log

**Comment by poettering** [comment]
lgtm to me, the big thing still missing is the serialization of the records. i.e. we either should return all response RRs, and that either in broken-down version (i.e. provide a resource_record_to_json() function that translates all fields to json counterparts, one by one) or a version that just serializes the binary rdata as base64.

Or best would probably be both, so that clients can rely on our parsers or use the binary versions, and they have freedom.

And I'd like to see a monitor client implemented.

**Comment by poettering** [comment]
(And I think "resolvectl monitor" when added should just work, i.e. enable the socket all the time. We allow dbus monitoring by default these days, i see no reason why we would be more restrictive for dns lookups there – as long as client is root it should be fine.)

(i mean, tcpdump doesn't need more than root privs, so why trying to be more catholic than the pope here, and require both explicit enablement of monitoring *and* access control)

File: src/resolve/resolved-dns-query.c (hunk 1 above)
Comment (poettering) [reply src/resolve/resolved-dns-query.c]: hmm, so we follow a cname chain on that request. Question is what we should return to clients here. Pretty sure we should return one varlink message for each element in the chain.

Comment (poettering) [reply src/resolve/resolved-dns-query.c]: so, basically, don't bother qith `q->request_packet`, it's never what you want. Aloway use `q->question_utf8`

**Comment by surajkrishnan14** [comment]
Thanks @bluca for taking care of these!

I've also started looking into `resource_record_to_json` to add support for more types, and plan to model it similar to `resource_record_to_string` - looks like we have the parsed data for all record types already in a union. I think this seems like a more usable option for clients compared to raw base64 encoded data. Or as @poettering says, it might be good to have both...

I agree that adding `resolvectl monitor` would also be useful.

**Pull request merged**